`docs:version` results in `Cannot find module './sidebars.json'`

<!--

  ! PLEASE HELP US HELP YOU !

  Bugs are fixed faster if you include:
  - a repro repository to inspect the code
  - an url to see the problem live (if possible)

  Pro tip: create a reproducible demo of the bug with https://new.docusaurus.io

-->

## 🐛 Bug Report

Trying to create a new version of the docs fails with an unhandled rejection

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yup

## To Reproduce

If possible, use https://new.docusaurus.io to create a CodeSandbox reproducible demo of the bug.

(Write your steps here:)

1. Clone https://github.com/facebook/jest
1. Run `yarn workspace jest-website docusaurus docs:version 27.0`

## Expected behavior

<!--
  How did you expect your project to behave?
  It’s fine if you’re not sure your understanding is correct.
  Write down what you thought would happen.
-->

It should correctly generate a new version of the docs

## Actual Behavior

<!--
  Did something go wrong?
  Is something broken, or not behaving as you expected?
  Describe this section in detail, and attach screenshots if possible.
  Don't only say "it doesn't work"!
-->

It errors out

```
(node:95658) UnhandledPromiseRejectionWarning: Error: Cannot find module './sidebars.json'
Require stack:
- /Users/simen/repos/jest/node_modules/@docusaurus/plugin-content-docs/lib/noop.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:885:15)
    at Function.Module._resolveFilename (/Users/simen/repos/jest/node_modules/module-alias/index.js:49:29)
    at resolveFileName (/Users/simen/repos/jest/node_modules/import-fresh/node_modules/resolve-from/index.js:29:39)
    at resolveFrom (/Users/simen/repos/jest/node_modules/import-fresh/node_modules/resolve-from/index.js:43:9)
    at module.exports (/Users/simen/repos/jest/node_modules/import-fresh/node_modules/resolve-from/index.js:46:41)
    at Object.module.exports [as default] (/Users/simen/repos/jest/node_modules/import-fresh/index.js:14:19)
    at Object.loadSidebars (/Users/simen/repos/jest/node_modules/@docusaurus/plugin-content-docs/lib/sidebars.js:130:47)
    at Object.cliDocsVersionCommand (/Users/simen/repos/jest/node_modules/@docusaurus/plugin-content-docs/lib/cli.js:64:43)
    at Command.<anonymous> (/Users/simen/repos/jest/node_modules/@docusaurus/plugin-content-docs/lib/index.js:56:23)
    at Command.listener [as _actionHandler] (/Users/simen/repos/jest/node_modules/@docusaurus/core/node_modules/commander/index.js:413:31)
    at Command._parseCommand (/Users/simen/repos/jest/node_modules/@docusaurus/core/node_modules/commander/index.js:914:14)
    at Command._dispatchSubcommand (/Users/simen/repos/jest/node_modules/@docusaurus/core/node_modules/commander/index.js:865:18)
    at Command._parseCommand (/Users/simen/repos/jest/node_modules/@docusaurus/core/node_modules/commander/index.js:882:12)
    at Command.parse (/Users/simen/repos/jest/node_modules/@docusaurus/core/node_modules/commander/index.js:717:10)
    at run (/Users/simen/repos/jest/node_modules/@docusaurus/core/bin/docusaurus.js:319:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:95658) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:95658) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Also note (as the warnings indicate) that the error is unhandled, meaning the CLI exits with status 0. You are missing a `catch` somewhere 🙂 

## Your Environment

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Public source code: https://github.com/facebook/jest/tree/master/website
- Public site url: https://jestjs.io/
- Docusaurus version used: `2.0.0-alpha.c32d8bd11`, but `2.0.0-beta.0` also fails with the same error
- Environment name and version (e.g. Chrome 78.0.3904.108, Node.js 10.17.0): Node 14.17.0
- Operating system and version (desktop or mobile): N/A

## Reproducible Demo

See reproduction steps above

--- a/packages/docusaurus-plugin-content-docs/src/cli.ts
+++ b/packages/docusaurus-plugin-content-docs/src/cli.ts
@@ -17,7 +17,7 @@ import {
   UnprocessedSidebarItem,
   UnprocessedSidebars,
 } from './types';
-import {loadSidebars} from './sidebars';
+import {loadSidebars, resolveSidebarPathOption} from './sidebars';
 import {DEFAULT_PLUGIN_ID} from '@docusaurus/core/lib/constants';
 
 function createVersionedSidebarFile({
@@ -154,17 +154,11 @@ export function cliDocsVersionCommand(
     throw new Error(`${pluginIdLogPrefix}There is no docs to version !`);
   }
 
-  // Create sidebar file.
-  const normalizedSidebarPath =
-    sidebarPath && !path.isAbsolute(sidebarPath)
-      ? path.join(siteDir, sidebarPath)
-      : sidebarPath;
-
   createVersionedSidebarFile({
     siteDir,
     pluginId,
     version,
-    sidebarPath: normalizedSidebarPath,
+    sidebarPath: resolveSidebarPathOption(siteDir, sidebarPath),
   });
 
   // Update versions.json file.

--- a/packages/docusaurus-plugin-content-docs/src/cli.ts
+++ b/packages/docusaurus-plugin-content-docs/src/cli.ts
@@ -145,6 +145,7 @@ export function cliDocsVersionCommand(
 
   // Copy docs files.
   const docsDir = path.join(siteDir, docsPath);
+
   if (fs.existsSync(docsDir) && fs.readdirSync(docsDir).length > 0) {
     const versionedDir = getVersionedDocsDirPath(siteDir, pluginId);
     const newVersionDir = path.join(versionedDir, `version-${version}`);
@@ -153,7 +154,18 @@ export function cliDocsVersionCommand(
     throw new Error(`${pluginIdLogPrefix}There is no docs to version !`);
   }
 
-  createVersionedSidebarFile({siteDir, pluginId, version, sidebarPath});
+  // Create sidebar file.
+  const normalizedSidebarPath =
+    sidebarPath && !path.isAbsolute(sidebarPath)
+      ? path.join(siteDir, sidebarPath)
+      : sidebarPath;
+
+  createVersionedSidebarFile({
+    siteDir,
+    pluginId,
+    version,
+    sidebarPath: normalizedSidebarPath,
+  });
 
   // Update versions.json file.
   versions.unshift(version);

--- a/packages/docusaurus-plugin-content-docs/src/sidebars.ts
+++ b/packages/docusaurus-plugin-content-docs/src/sidebars.ts
@@ -25,11 +25,13 @@ import {
   SidebarItemsGeneratorVersion,
   NumberPrefixParser,
   SidebarItemsGeneratorOption,
+  PluginOptions,
 } from './types';
 import {mapValues, flatten, flatMap, difference, pick, memoize} from 'lodash';
 import {getElementsAround} from '@docusaurus/utils';
 import combinePromises from 'combine-promises';
 import {DefaultSidebarItemsGenerator} from './sidebarItemsGenerator';
+import path from 'path';
 
 type SidebarItemCategoryJSON = SidebarItemBase & {
   type: 'category';
@@ -256,7 +258,19 @@ export const DefaultSidebars: UnprocessedSidebars = {
 
 export const DisabledSidebars: UnprocessedSidebars = {};
 
+// If a path is provided, make it absolute
+// use this before loadSidebars()
+export function resolveSidebarPathOption(
+  siteDir: string,
+  sidebarPathOption: PluginOptions['sidebarPath'],
+): PluginOptions['sidebarPath'] {
+  return sidebarPathOption
+    ? path.resolve(siteDir, sidebarPathOption)
+    : sidebarPathOption;
+}
+
 // TODO refactor: make async
+// Note: sidebarFilePath must be absolute, use resolveSidebarPathOption
 export function loadSidebars(
   sidebarFilePath: string | false | undefined,
 ): UnprocessedSidebars {
@@ -279,6 +293,7 @@ export function loadSidebars(
 
   // We don't want sidebars to be cached because of hot reloading.
   const sidebarJson = importFresh(sidebarFilePath) as SidebarsJSON;
+
   return normalizeSidebars(sidebarJson);
 }
 

--- a/packages/docusaurus-plugin-content-docs/src/versions.ts
+++ b/packages/docusaurus-plugin-content-docs/src/versions.ts
@@ -24,6 +24,7 @@ import {DEFAULT_PLUGIN_ID} from '@docusaurus/core/lib/constants';
 import {LoadContext} from '@docusaurus/types';
 import {getPluginI18nPath, normalizeUrl, posixPath} from '@docusaurus/utils';
 import {difference} from 'lodash';
+import {resolveSidebarPathOption} from './sidebars';
 
 // retro-compatibility: no prefix for the default plugin id
 function addPluginIdPrefix(fileOrDir: string, pluginId: string): string {
@@ -184,9 +185,7 @@ function getVersionMetadataPaths({
 
   function getSidebarFilePath() {
     if (isCurrentVersion) {
-      return options.sidebarPath
-        ? path.resolve(context.siteDir, options.sidebarPath)
-        : options.sidebarPath;
+      return resolveSidebarPathOption(context.siteDir, options.sidebarPath);
     } else {
       return path.join(
         getVersionedSidebarsDirPath(context.siteDir, options.id),

--- a/website/docusaurus.config.js
+++ b/website/docusaurus.config.js
@@ -224,7 +224,7 @@ const isVersioningDisabled = !!process.env.DISABLE_VERSIONING || isI18nStaging;
         docs: {
           // routeBasePath: '/',
           path: 'docs',
-          sidebarPath: require.resolve('./sidebars.js'),
+          sidebarPath: 'sidebars.js',
           editUrl: ({locale, docPath}) => {
             if (locale !== 'en') {
               return `https://crowdin.com/project/docusaurus-v2/${locale}`;

